Client/Experiments/Messaging/GleanPlumbContextProvider.swift

```diff
@@ -40,18 +40,18 @@ class GleanPlumbContextProvider {
 
     var isInactiveNewUser: Bool {
         // existing users don't have firstAppUse set
-        guard let firstAppUse = userDefaults.object(forKey: PrefsKeys.KeyFirstAppUse) as? UInt64
+        guard let firstAppUse = userDefaults.object(forKey: PrefsKeys.KeyFirstAppUse) as? UInt64,
+              let lastSession = userDefaults.object(forKey: PrefsKeys.KeyLastSession) as? UInt64
         else { return false }
 
         let now = Date()
         let notificationDate = Date.fromTimestamp(firstAppUse + Constant.activityReferencePeriod)
 
-        // check that we are not past the reference time for inactive user check
-        guard now < notificationDate else { return false }
-
         // We don't care how often the user is active in the first 24 hours after first use.
         // If they are not active in the second 24 hours after first use they are considered inactive.
-        return now < Date.fromTimestamp(firstAppUse + Constant.inactivityPeriod)
+        let lastSessionDate = Date.fromTimestamp(lastSession)
+        return now >= Date.fromTimestamp(firstAppUse + Constant.activityReferencePeriod) &&
+            lastSessionDate <= Date.fromTimestamp(firstAppUse + Constant.inactivityPeriod)
```

Comment:
@jhugman I had explored this and dismissed it for two reasons:

1. JEXL was/is not testable so verifying the trigger is correct can only be done manually which could lead to the user getting the message at the wrong time or not at all
2. I haven't found a way to make JEXL represent this exact scenario. We don't know when exactly Apple will run the background task so let's say we get run 50h after install. With JEXL we only check the last 24h for activity so if the user was active 25h after install we'd miss that. The requirement said the user should be inactive for the second 24h after install so technically JEXL would trigger at the wrong time. Additionally I think we use the install time instead of first app use time to determine the time range for the 48h after install. Now if the user installs the app but doesn't open it for another 10h we'd have another false positive with the JEXL trigger.

Please correct me if any of the above is wrong.

Reply:
> JEXL was/is not testable so verifying the trigger is correct can only be done manually which could lead to the user getting the message at the wrong time or not at all

Ok; the testing APIs really start coming through in [v97.3.0](https://github.com/mozilla/application-services/releases/tag/v97.3.0).

The `Hours` counter saves a rolling 72 hours, but let's see if we can work out how to use the event-store for next time. :)